File: leetcode/145.binary_tree_postorder_traversal.rb
'''

Given a binary tree, return the postorder traversal of its nodes\' values.

For example:
Given binary tree [1,null,2,3],

   1
    \
     2
    /
   3
 

return [3,2,1].
'''

# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

# @param {TreeNode} root
# @return {Integer[]}
def postorder_traversal(root)
  return [] if root.nil? || (root.left.nil? && root.right.nil?)
  postorder_traversal(root.left) + postorder_traversal(root.right) + [root.val]
end

File: leetcode/pythonSolutions/145.binary_tree_postorder_traversal.py
'''
145. Binary Tree Postorder Traversal
Hard

Given a binary tree, return the postorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
    \
     2
    /
   3

Output: [3,2,1]
Follow up: Recursive solution is trivial, could you do it iteratively?
'''

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        if root:
          
            if root.left:
              result = result + self.postorderTraversal(root.left)
  
            if root.right:
              result = result + self.postorderTraversal(root.right)
            result.append(root.val)
        return result